### PR TITLE
feat: add params for single select v1.2

### PR DESCRIPTION
**Pull request by allisonc07**
SDK changes for single select v1.2:

1. Update  README so that `single_select` can be a bool or a dictionary
2. Add ability to add `single_select_vin` to auth url
3. Update tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -110,25 +110,26 @@ Checkout our [Errors documentation][errors] to learn more.
...
-| `scope`         | String[] |**Optional** List of permissions your application requires. This will default to requiring all scopes. The valid permission names are found in the [API Reference](https://smartcar.com/docs/api#get-all-vehicles). |
-| `test_mode`   | Boolean |**Optional** Launch the Smartcar Connect in test mode. |
+| Parameter       | Type | Required | Description   |
+|:--------------- |:---|:---|:------------- |
+| `client_id`     | String |**Required** | Application clientId obtained from [Smartcar Developer Portal](https://dashboard.smartcar.com). |
+| `client_secret` | String |**Required** | Application clientSecret obtained from [Smartcar Developer Portal](https://dashboard.smartcar.com). |
+| `redirect_uri`  | String |**Required** | RedirectURI set in [application settings](https://dashboard.smartcar.com/apps). Given URL must match URL in application settings. |
+| `scope`         | String[] |**Optional** | List of permissions your application requires. This will default to requiring all scopes. The valid permission names are found in the [API Reference](https://smartcar.com/docs/api#get-all-vehicles). |
```
Comment (charlottekosche) [inline README.md]: @jerzzhang, does the scope default to "required"?

Comment (charlottekosche) [reply README.md]: We need to adapt this copy once the Scope Guide is finalized.

File: README.md [hunk 2]
```diff
@@ -110,25 +110,26 @@ Checkout our [Errors documentation][errors] to learn more.
...
+| `test_mode`   | Boolean |**Optional** | Launch the Smartcar Connect in test mode. |
```
Comment (charlottekosche) [inline README.md]: Launch the Smartcar Connect **flow** in test mode.

File: README.md [hunk 3]
```diff
@@ -110,25 +110,26 @@ Checkout our [Errors documentation][errors] to learn more.
...
 #### Arguments
-| Parameter       | Type | Description   |
-|:--------------- |:---|:------------- |
-| `force`   | Boolean |**Optional** Setting `forcePrompt` to `true` will show the permissions approval screen on every authentication attempt, even if the user has previously consented to the exact scope of permissions. |
-| `state`         | String |**Optional** OAuth state parameter passed to the redirectUri. This parameter may be used for identifying the user who initiated the request. |
-| `vehicle_info['make']`  | String |**Optional** Including the dict `vehicle_info` with a `make` property allows users to bypass the car brand selection screen. For a complete list of supported makes, please see our [API Reference](https://smartcar.com/docs/api#authorization) documentation. Makes are case-insensitive. |
-
+| Parameter       | Type | Required | Description   |
+|:--------------- |:---|:---|:------------- |
+| `force`   | Boolean |**Optional** | Setting `forcePrompt` to `true` will show the permissions approval screen on every authentication attempt, even if the user has previously consented to the exact scope of permissions. |
```
Comment (charlottekosche) [inline README.md]: Setting `forcePrompt` to `true` will show the permissions approval screen on every **authorization** attempt, even if the user has previously consented to the exact scope of permissions.

File: README.md [hunk 4]
```diff
@@ -110,25 +110,26 @@ Checkout our [Errors documentation][errors] to learn more.
...
+| `state`         | String |**Optional** | OAuth state parameter passed to the redirectUri. This parameter may be used for identifying the user who initiated the request. |
```
Comment (charlottekosche) [inline README.md]: OAuth state parameter passed to the **redirect URI**. This parameter may be used **to identify** the user who **has** initiated the request.

File: README.md [hunk 5]
```diff
@@ -110,25 +110,26 @@ Checkout our [Errors documentation][errors] to learn more.
...
+| `vehicle_info['make']`  | String |**Optional** | Including the dict `vehicle_info` with a `make` property allows users to bypass the car brand selection screen. For a complete list of supported makes, please see our [API Reference](https://smartcar.com/docs/api#authorization) documentation. Makes are case-insensitive. |
```
Comment (charlottekosche) [inline README.md]: leave out "documentation"
**The names of makes** are case-insensitive.

File: README.md [hunk 6]
```diff
@@ -110,25 +110,26 @@ Checkout our [Errors documentation][errors] to learn more.
...
+| `single_select` | Boolean or Dictionary | **Optional** | An optional value that sets the behavior of the grant dialog displayed to the user. If set to `true`, `single_select` limits the user to selecting only one vehicle. Defaults to `false`. See the [Single Select guide](https://smartcar.com/docs/guides/single-select/) for more information.
```
Comment (charlottekosche) [inline README.md]: This need to be consistent with whatever it says in the API reference. There's still an open PR for that, so whatever the copy will be in the end, we need to adjust here 😊

File: README.md [hunk 7]
```diff
@@ -110,25 +110,26 @@ Checkout our [Errors documentation][errors] to learn more.
...
+| `single_select.vin` | String | **Optional** | The VIN (Vehicle Identifier Number) of a vehicle. By passing in `single_select` as a dict with a `vin` property, the single select behavior will be turned on and Smartcar will authorize only the vehicle with the specified VIN. The `single_select_vin` takes precedence over the `make` parameter. Please refer to the [Single Select guide](https://smartcar.com/docs/guides/single-select/).
```
Comment (charlottekosche) [inline README.md]: This looks perfect, apart from the first sentence. Could we replace the first sentence with something like:
**An optional value that sets the behavior of the grant dialog displayed to the user. The `single-select` parameter takes a VIN (vehicle identification number) as a property.**

File: README.md [hunk 8]
```diff
@@ -190,10 +191,10 @@ A compatible vehicle is a vehicle that:
...
 
 #### Arguments
-| Parameter       | Type | Description   |
-|:--------------- |:---|:------------- |
-| `vin`         | String |The VIN of the vehicle. |
-| `scope`       | String[] | The list of permissions to check compatibility for. Valid permission names are found in the [API Reference](https://smartcar.com/docs/api#get-all-vehicles).
+| Parameter       | Type | Required | Description   |
+|:--------------- |:---|:---|:------------- |
+| `vin`         | String | **Required** | The VIN of the vehicle. |
+| `scope`       | String[] | **Required** | The list of permissions to check compatibility for. Valid permission names are found in the [API Reference](https://smartcar.com/docs/api#get-all-vehicles).
```
Comment (charlottekosche) [inline README.md]: Valid permission names **can be** found in the [API Reference]

File: README.md [hunk 9]
```diff
@@ -349,11 +350,11 @@ Check if an expiration is expired
...
-| Parameter       | Type | Description   |
-|:--------------- |:---- |:------------- |
-| `access_token`    | String | **Required** A valid access token from a previously retrieved access object |
-| `limit`    | Integer | **Optional** The number of vehicle ids to return |
-| `offset`    | Integer | **Optional** The index to start the vehicle list at |
+| Parameter       | Type | Required | Description   |
+|:--------------- |:---|:---|:------------- |
+| `access_token`    | String | **Required** | A valid access token from a previously retrieved access object |
+| `limit`    | Integer | **Optional** | The number of vehicle ids to return |
+| `offset`    | Integer | **Optional** | The index to start the vehicle list at |
```
Comment (charlottekosche) [inline README.md]: The index **at which the vehicle list should start**

File: smartcar/smartcar.py [hunk 10]
```diff
@@ -97,7 +97,14 @@ def get_auth_url(self, force=False, state=None, vehicle_info=None):
             vehicle_info (dict, optional): A dict with a property, make. Allows
                 users to bypass the car brand selection screen, allowing the
                 user to go directly to the vehicle login screen. 
-                Defaults to None. 
+                Defaults to None.
+            single_select (bool or dictionary, optional): An optional value that
```
Comment (charlottekosche) [inline smartcar/smartcar.py]: single_select (**boolean** or dictionary, optional):...

File: smartcar/smartcar.py [hunk 11]
```diff
@@ -85,7 +85,7 @@ def __init__(self, client_id, client_secret, redirect_uri, scope=None, test_mode
         else:
             self.test_mode = test_mode if test_mode else False
 
-    def get_auth_url(self, force=False, state=None, vehicle_info=None):
+    def get_auth_url(self, force=False, state=None, vehicle_info=None, single_select=False):
```
Comment (adolfoportilla) [inline smartcar/smartcar.py]: should default to none right?

File: smartcar/smartcar.py [hunk 12]
```diff
@@ -97,7 +97,14 @@ def get_auth_url(self, force=False, state=None, vehicle_info=None):
...
                 user to go directly to the vehicle login screen. 
-                Defaults to None. 
+                Defaults to None.
+            single_select (bool or dictionary, optional): An optional value that
+                sets the behavior of the grant dialog displayed to the user. It
+                can be either a bool or dict. If set to True, `single_select` 
+                limits the user to selecting only one vehicle. If `single_select`
+                is a dictionary with the property `vin`, Smartcar will only authorize the vehicle
+                with the specified VIN. See the [Single Select guide](https://smartcar.com/docs/guides/single-select/)
+                for more information. Defaults to False.
```
Comment (adolfoportilla) [inline smartcar/smartcar.py]: defaults to None right?

File: tests/test_smartcar.py [hunk 13]
```diff
@@ -189,7 +189,56 @@ def test_get_auth_url_vehicle_info_dictionary(self):
...
 
         expected = smartcar.const.CONNECT_URL + '/oauth/authorize?' + query
-        
+
+        expected_params = parse_qs(expected)
+        actual_params = parse_qs(actual)
+
+        assertDeepEquals(self, expected_params, actual_params)
+
+    def test_get_auth_url_single_select_bool(self):
```
Comment (adolfoportilla) [inline tests/test_smartcar.py]: Add test for without single select?

File: tests/test_smartcar.py [hunk 14]
```diff
@@ -189,7 +189,56 @@ def test_get_auth_url_vehicle_info_dictionary(self):
...
+
+        actual = client.get_auth_url(force=True, state='stuff', single_select=single_select)
+
+        query = urlencode({
+            'response_type': 'code',
+            'client_id': self.client_id,
+            'redirect_uri': self.redirect_uri,
+            'approval_prompt': 'force',
+            'state': 'stuff',
+            'scope': ' '.join(self.scope),
```
Comment (adolfoportilla) [inline tests/test_smartcar.py]: there should also be single_select here in the query right?

File: smartcar/smartcar.py (hunk 11 above)
Comment (allisonc07) [reply smartcar/smartcar.py]: yup, I think I didn't catch this when I rebased! fixed!

File: smartcar/smartcar.py (hunk 12 above)
Comment (allisonc07) [reply smartcar/smartcar.py]: got it!

File: tests/test_smartcar.py (hunk 13 above)
Comment (allisonc07) [reply tests/test_smartcar.py]: :+1:

File: tests/test_smartcar.py (hunk 14 above)
Comment (allisonc07) [reply tests/test_smartcar.py]: yes!`

File: tests/test_smartcar.py (hunk 13 above)
Comment (allisonc07) [reply tests/test_smartcar.py]: actually didn't add this one because it is implicitly tested in our default getAuthUrl

**Pull request merged**